### ISSUE TITLE
The results of QuantizationSimModel and DSP deployment are different

### ISSUE THREAD

**Issue by hammer422** [quic/aimet#1303]
**Definition of my model**
`class Net(nn.Module):

    def __init__(self):
        super(Net, self).__init__()
        self.avg_pool = nn.AdaptiveAvgPool2d(1)
        self.fc1 = nn.Linear(3, 256)
        self.fc2 = nn.Linear(256, 10)

    def forward(self, x): # (1,3,20,20)
        x = self.avg_pool(x)
        x = x.view(x.size(0), -1)
        x = self.fc1(x)
        x = self.fc2(x)
        return x`

**My quantization process is**

**_apply_CLE -> apply_bias_correction  -> calc_quantsim_error -> export_quantsim_**

`   def calc_quantsim_error(model: nn.Module, evalutor):

        dummy_input = torch.rand((1,3,20,20))
        iterations = RAW_DATA_LIST_LEN // BATCH_SIZE

        quantsim = QuantizationSimModel(model=model, quant_scheme='tf_enhanced',
                                        dummy_input=dummy_input, rounding_mode='nearest',
                                        default_output_bw=8, default_param_bw=8, in_place=False)

        quantsim.compute_encodings(forward_pass_callback=partial(evalutor),
                                   forward_pass_callback_args=iterations)

        return evalutor(quantsim.model)
`

`    def export_quantsim(model:nn.Module, evalutor):

        dummy_input = torch.rand((1,3,20,20))
        iterations = RAW_DATA_LIST_LEN // BATCH_SIZE

        quantsim = QuantizationSimModel(model=model, quant_scheme='tf_enhanced',
                                        dummy_input=dummy_input, rounding_mode='nearest',
                                        default_output_bw=8, default_param_bw=8, in_place=False)
        quantsim.compute_encodings(forward_pass_callback=partial(evalutor),
                                   forward_pass_callback_args=iterations)

        from aimet_torch.quantsim import OnnxExportApiArgs

        onnx_params = OnnxExportApiArgs(opset_version=None, input_names=["inp"], output_names=["logits"])
        quantsim.export("./aimet_quat/", "tmp", dummy_input, onnx_params)
        return evalutor(quantsim.model, show=True)
`

**My SNPE process is** 
`./snpe-onnx-to-dlc --input_network tmp.onnx --output_path tmp.dlc --quantization_overrides tmp.encodings`
`./snpe-dlc-quantize --input_dlc tmp.dlc --input_list tmp_file.txt --output_dlc tmp_quat_aimet.dlc --override_params`

**Result:**
My input is always a matrix(1,3,20,20) with all 0.5, like np.ones((1,3,20,20)) * 0.5

Pytorch original float32 model:
tensor([ 0.3262,  0.5271,  0.2023,  0.2009, -0.1790, -0.1427, -0.3627,  0.4660,
        -0.1745,  0.0697])

quantsim.model:
show result 
[ 0.32619095  0.527073    0.20226319  0.20089692 -0.17902157 -0.1427542
 -0.36271656  0.46603447 -0.17453653  0.06965145]

SNPE run on DSP(tmp_quat_aimet.dlc):
logits
0.104272  0.371904  0.0278059  0.0417089  -0.0868935  -0.361477  -0.250253  0.521361  0.111224  -0.166836

Why there is such a big gap between QuantizationSimModel and SNPE DSP?

Did I do anything wrong?

**Comment by quic-ssiddego** [comment]
@changewOw Your CLE updates look fine. Could you update the command to include --use_encoding_optimizations flag with (dlc quantize command)

**Comment by hammer422** [comment]
Thank you. I found that the such a **big difference** is caused by DSP. I don't think it has anything to do with AIMET. There seems to be a bug with the SNPE Linear layer. My snpe version is snpe-1.61.0.3358

I try to include --use_encoding_optimizations and no change in results.


I want to show another case.
```

class Net(nn.Module):
    def __init__(self, show=False):
        super(Net, self).__init__()
        self.avg_pool = nn.AdaptiveAvgPool2d(1)
        self.conv = nn.Conv2d(3, 40,1,1, bias=False)
        self.conv1 = nn.Conv2d(40, 20,1,1,bias=False)
        self.relu = nn.ReLU()
    def forward(self, x, debug=False): # (1,3,20,20)
        x = self.avg_pool(x)
        print(x)
        x = self.conv(x)
        print(x)
        x = self.conv1(x)
        print(x)
        x = torch.flatten(x, 1)
        print(x)
        return x
```

I don't do CLE or BC or any other Quant tech. I directly do model export.


**Result:**
**quantsim.model:**
[ 0.059581  0.037238  0.018619 -0.14523  -0.402175 -0.014895  0.044686
  0.011172  0.059581 -0.171297 -0.014895  0.227154 -0.22343   0.201087
  0.115439  0.27184  -0.070753  0.040962 -0.387279  0.152677]

**SNPE run on DSP(tmp_quat_aimet.dlc):**
logits.raw
[ 0.059581  0.037238  0.018619 **_-0.148954_** **_-0.398451_** -0.014895  0.044686
  0.011172  0.059581 -0.171297 -0.011172  0.227154 -0.22343   0.201087
  0.115439  0.27184  -0.070753  0.040962 -0.387279  0.152677]


The result seems not to be bit exact.When I have more and more convolution layers, the error will be larger

Here is my dlc convert command
`./snpe-onnx-to-dlc --input_network tmp.onnx --output_path tmp.dlc --quantization_overrides tmp.encodings
`
`./snpe-dlc-quantize --input_dlc tmp.dlc --input_list tmp_file.txt --output_dlc tmp_quat_aimet.dlc --override_params --debug3 --bias_bitwidth 32`
I also try to include --use_encoding_optimizations and no change in results.

@quic-ssiddego I think AIMET&SNPE dsp should be BIT exact. What caused the above problem?





I put the results of the middle layer here

QuantSim Model:
tensor([0.701001, 0.701001, 0.701001], device='cuda:0')
tensor([ 0.546874,  0.367571,  0.394466,  0.134477,  0.546874,  0.313780,
        -0.268954,  0.053791, -0.618595, -0.600665,  0.331710,  0.179303,
         0.430327, -0.125512,  0.762037,  0.125512, -0.439292, -0.134477,
         0.412397,  0.026895, -0.125512, -0.098617, -0.143442,  0.259989,
        -0.340676, -0.421362,  0.277920, -0.224129,  0.376536, -0.394466,
         0.224129, -0.233094,  0.089651, -0.519978, -0.340676, -0.842724,
        **_-0.215163_**, -0.179303, -0.340676, -0.771003], device='cuda:0')
tensor([ 0.059581,  0.037238,  0.018619, -0.145230, -0.402175, -0.014895,
         0.044686,  0.011172,  0.059581, -0.171297, -0.014895,  0.227154,
        -0.223430,  0.201087,  0.115439,  0.271840, -0.070753,  0.040962,
        -0.387279,  0.152677], device='cuda:0')
tensor([ 0.059581,  0.037238,  0.018619, **_-0.145230, -0.402175_**, -0.014895,
         0.044686,  0.011172,  0.059581, -0.171297, **_-0.014895,_**  0.227154,
        -0.223430,  0.201087,  0.115439,  0.271840, -0.070753,  0.040962,
        -0.387279,  0.152677], device='cuda:0')
show target 
[ 0.061701  0.03675   0.018051 -0.147992 -0.397405 -0.015342  0.045102
  0.01008   0.05833  -0.17144  -0.015069  0.225478 -0.227847  0.199583
  0.115799  0.267447 -0.071691  0.039989 -0.388088  0.155455]
show result 
[ 0.059581  0.037238  0.018619 -0.14523  -0.402175 -0.014895  0.044686
  0.011172  0.059581 -0.171297 -0.014895  0.227154 -0.22343   0.201087
  0.115439  0.27184  -0.070753  0.040962 -0.387279  0.152677]



SNPE DSP:
##################################################
4.raw
[0.7 0.7 0.7]
##################################################
7.raw
[ 0.546874  0.367571  0.394466  0.134477  0.546874  0.31378  -0.268954
  0.053791 -0.618595 -0.600665  0.33171   0.179303  0.430327 -0.125512
  0.762037  0.125512 -0.439292 -0.134477  0.412397  0.026895 -0.125512
 -0.098617 -0.143442  0.259989 -0.340676 -0.421362  0.27792  -0.224129
  0.376536 -0.394466  0.224129 -0.233094  0.089651 -0.519978 -0.340676
 -0.842724 **_-0.206198_** -0.179303 -0.340676 -0.771003]
##################################################
10.raw
[ 0.059581  0.037238  0.018619 **_-0.148954 -0.398451_** -0.014895  0.044686
  0.011172  0.059581 -0.171297 **_-0.011172_**  0.227154 -0.22343   0.201087
  0.115439  0.27184  -0.070753  0.040962 -0.387279  0.152677]
##################################################
10.ncs.raw
[ 0.059581  0.037238  0.018619 -0.148954 -0.398451 -0.014895  0.044686
  0.011172  0.059581 -0.171297 -0.011172  0.227154 -0.22343   0.201087
  0.115439  0.27184  -0.070753  0.040962 -0.387279  0.152677]

**Comment by hammer422** [comment]
Update ResNet18 example:

```
class resnet18_removefc(nn.Module):
    def __init__(self):
        super().__init__()
        self.model = resnet18(pretrained=True)
    def forward(self, x):
        x = self.model.conv1(x)
        x = self.model.bn1(x)
        x = self.model.relu(x)
        x = self.model.maxpool(x)
        x = self.model.layer1(x)
        x = self.model.layer2(x)
        x = self.model.layer3(x)
        x = self.model.layer4(x)
        x = self.model.avgpool(x)
        x = torch.flatten(x, 1)
        return x

dummy_input = np.zeros((1,3,224,224), dtype=np.float32)
dummy_input_torch = torch.from_numpy(dummy_input).float()
dummy_input_torch = dummy_input_torch.cuda()

def export_quantsim(model:nn.Module)

    def evalutor(model:nn.Module, dummy_input:torch.Tensor):
        model.eval()

        with torch.no_grad():
            _ = model(dummy_input)

    quantsim = QuantizationSimModel(model=model, quant_scheme="tf_enhanced",
                                dummy_input=dummy_input_torch, rounding_mode='nearest',
                                default_output_bw=8, default_param_bw=8, in_place=False)

    quantsim.compute_encodings(forward_pass_callback=partial(evalutor),
                                forward_pass_callback_args=(dummy_input_torch))

    from aimet_torch.quantsim import OnnxExportApiArgs

    onnx_params = OnnxExportApiArgs(opset_version=None, input_names=["image"], output_names=["logits"])
    quantsim.export("../aimet_quat/tmp", "tmp", dummy_input_torch.cpu(), onnx_params)

    return quantsim

model = resnet18_removefc()
model.eval()
model.cuda()

print("pytorch org fp32 output")
print(model(dummy_input_torch)[0, :20])

quantsim = export_quantsim(model)

print("quantsim output")
print(quantsim.model(dummy_input_torch)[0, :20])
```



**_Here is the result_**
pytorch org fp32 output
tensor([0.006306, 0.000000, 0.010852, 0.209028, 0.000000, 0.000884, 0.064607,
        0.166484, 0.444869, 0.052149, 0.043465, 0.021005, 0.029137, 0.285485,
        0.000000, 0.242928, 0.000000, 0.386071, 0.030778, 0.000000],
       device='cuda:0', grad_fn=<SliceBackward>)


quantsim output
tensor([0.000000, 0.000000, 0.017582, 0.316479, 0.000000, 0.000000, 0.035164,
        0.123075, 0.351643, 0.123075, 0.035164, 0.070329, 0.070329, 0.175821,
        0.000000, 0.263732, 0.000000, 0.334061, 0.017582, 0.000000],
       device='cuda:0', grad_fn=<SliceBackward>)


SNPE DSP 
./snpe-net-run --container /sdcard/tmp_quat_aimet.dlc --input_list /sdcard/tmp_file_android.txt --use_dsp
logits.raw
[1.075786 0.117664 1.428778 0.890885 0.840458 0.20171  0.20171  0.453847
 0.907694 0.806839 0.302565 1.143022 0.974931 0.537893 0.151282 1.428778
 1.294305 1.697724 1.58006  0.470656 1.680915 1.966671 0.168092 0.504275



You can see that the results of quantsim and DSP are inconsistent

**Comment by Stephenfang51** [comment]
@changewOw 你好 请问你找到了可能的原因了吗？

**Comment by hammer422** [comment]
Sorry, it's been a long time. I don't remember the exact results at that time. 
The conclusion is that 
1. there is a problem with SNPE linear layer. 
2. SNPE and quic are not bit aligned

**Comment by AlmogDavid** [comment]
Why this issue is closed?! it's an open bug that was not solved.
